Slur Relation & Space Drag

yinanazhou opened the pull request:
- #258: Relation visualization improvements
  - Implemented slur-based visualization
  - Added support for multiple slurs
- #259: Layer visibility improvements
  - Fixed cropped elements visibility
- #263: New map and navigate with space drag
  - Fixed view width issues in Safari & Firefox

yinanazhou committed:
fix: use slur for relation instead of rounded hull

- Added `draw_slur()`
- Adjusted style for relation

refs: #258

yinanazhou committed:
fix: handle multiple slurs

- Added `count_existing_slurs()`
- Adjusted offset for multiple slurs

refs: #258

yinanazhou committed:
style: prevent layers overlap and add top border

yinanazhou committed:
style: make cropped elements in layers visible

refs: #259

yinanazhou committed:
fix: initialize `current_draw_context` on load

yinanazhou committed:
fix: adjust svg dimension on relation changes

- Added `adjustSvgDimensions`

yinanazhou committed:
fix: make `svg_container` an SVG element

yinanazhou committed:
feat: create new space drag for score navigation

yinanazhou committed:
fix: remove minimap

yinanazhou committed:
style: fix view width in safari & firefox

refs: https://github.com/DCMLab/reductive_analysis_app/issues/263#issuecomment-2772527936

yrammos commented:
@yinanazhou a few observations:

- The PR seems too long to me. It would be best to dedicate a separate commit to the new visualization model (i.e. the use of slurs instead of highlights).
- I’m reluctant to approve a PR that crashes on Safari.
- Most importantly, the slurs-based model is problematic with long scores: scores appear illegibly tiny and, while zooming in could alleviate the problem, zoomed-in scores are not scrollable (at least on Firefox). I am going to send you an example MEI file via Mattermost.

yrammos commented:
I think we’re on the right track. But we need to make sure the new model works with scores that are large (either in duration, or in number of staves).

yinanazhou commented:
Hi @yrammos,

> * The PR seems too long to me. It would be best to dedicate a separate commit to the new visualization model (i.e. the use of slurs instead of highlights).

Thanks for the suggestion—I'll keep that in mind for future PRs. That said, smaller PRs also depend on timely review to stay manageable and avoid merge conflicts.

> * I’m reluctant to approve a PR that crashes on Safari.

As noted [here](https://github.com/DCMLab/reductive_analysis_app/issues/263#issuecomment-2773342187), I haven’t been able to reproduce the crash on Safari. I’ve attached a screenshot—could you please try again?  
<img width="1286" alt="image" src="https://github.com/user-attachments/assets/646bbd9e-01c9-4ec4-9853-eb9f1ba71e77" />

> * Most importantly, the slurs-based model is problematic with long scores: scores appear illegibly tiny and, while zooming in could alleviate the problem, zoomed-in scores are not scrollable (at least on Firefox). I am going to send you an example MEI file via Mattermost.

Thanks for the file. Here’s a screenshot of it opened in Firefox:  
<img width="1396" alt="image" src="https://github.com/user-attachments/assets/5f63332f-97f1-45c1-8d1e-43094ccd982c" />

I’m able to drag the score without issues. I agree that navigating large scores with zoom-in buttons is not ideal at the moment—please open a separate issue for improving the UI.

yrammos commented:
@yinanazhou my Safari version is 18.4 (20621.1.15.11.10). There are two persistent issues on my Safari:

- An `Unhandled Promise Rejection: EncodingError: Decoding failed` error on startup (even before attempting to load a score).
- A `TypeError: undefined is not an object (evaluating 'midiPlayer.stop')` error when attempting to load a score. In general, midiPlayer seems not to be instantiated (none of the `midiPlayer` methods works).

This could suggest an issue with the `this.instrument.load.then(…)` block on line 178 of `src/js/new/modules/Player/index.js`.

Which version of node are you using? I am running on 18.12.0.

yrammos commented:
We can treat the score sizing issue as a separate one, if you prefer.

Before we do so, though, I’d like to better understand why the size of the Verovio rendering depends on score length in this new implementation. In the original version, scores had a fixed note size on startup, and there was generally no need to zoom in/out before annotating. What is your reason for scaling the score on startup? 

(By the way, the rendering of the long example file on my Firefox is much worse than it appears on your screenshot. Here is what I get. This MEI file does not include any annotations, so the rendering should not be resized by `adjustSvgDimensions`, right?)

<img width="1792" alt="Screenshot 2025-04-04 at 10 55 49" src="https://github.com/user-attachments/assets/1cebaedd-a6bc-4b59-97fa-89d2e02e7fdd" />

In general, it may make sense to rethink layer rendering (and zoom), so that all (meta-)relations are always visible at any zoom level, and zooming in/out resizes the entire layer, including the containing div itself.

yrammos commented on src/js/ui.js:
```diff
@@ -409,14 +422,11 @@ export function drag_selector_installer(svg_elem) {
     ) {
       if ($('.ds-selector').height() > 10 || $('.ds-selector').width() > 10) {
-        $('#minimap').fadeOut(300)
         // Icon credit: pixel-perfect (flaticon.com).
```
Is the FlatIcon license permissive enough for us to include this icon?

yrammos replied:
Oh dear. According to [this](https://support.flaticon.com/s/article/Attribution-How-when-and-where-FI?language=en_US) article, we should add an attribution.

> Insert the attribution on the page where the icon is shown. This can be placed next to the image or on the footer of the website.

This seems overkill, though, unless done very discreetly (e.g. light-gray fine print on bottom-right corner).

yrammos replied:
Instead of a base64 graphic, could we not use simple CSS for the hand? I’m thinking of something like `cursor: grab`.

yinanazhou replied:
I agree it's something we should look into. Just to clarify, this code wasn’t added by me and is unrelated to the scope of this PR. I’d suggest opening a separate issue so we can address it properly.

yinanazhou commented:
I'm using Safari 18.3.1 (20620.2.4.11.6) with Node 20.18.1.

The error message you're encountering appears to be related to the player loading and doesn't seem to be caused by the changes introduced in this PR.

I suggest opening a separate issue for this, and I'll be happy to address it in a follow-up PR.

yinanazhou committed:
style: update Verovio options and container dimensions

- Adjusted Verovio rendering options
- Updated container dimensions and visibilities

refs: #https://github.com/DCMLab/reductive_analysis_app/pull/264#issuecomment-2777847009

yrammos commented:
We can merge this in.

The Safari issues have been located within an unrelated commit (see #267).

The new UIs (slurs and space-dragging) both have rough edges but these can be addressed in separate tickets with small, granular PRs.
